File: go/0049.go
package leetcode

func groupAnagrams(strs []string) [][]string {
	wordToGroup := make(map[[26]byte][]string, len(strs))

	for _, word := range strs {
		var letterFrequency [26]byte
		for _, letter := range word {
			letterFrequency[letter-'a']++
		}
		wordToGroup[letterFrequency] = append(wordToGroup[letterFrequency], word)
	}

	anagramGroups := make([][]string, len(wordToGroup))
	idx := 0
	for _, wordList := range wordToGroup {
		anagramGroups[idx] = wordList
		idx++
	}

	return anagramGroups
}

File: python/0049.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)
        for word in strs:
            letter_counts = [0] * 26
            for letter in word:
                letter_counts[ord(letter) - ord("a")] += 1
            groups[tuple(letter_counts)].append(word)
        return groups.values()